packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -274,16 +286,82 @@ const filterVerifiedProposals = async (
   });
 };
 
+// Filters out all proposals that have been executed on-chain. This results in proposals both before and after their
+// challenge period.
+const filterUnexecutedProposals = async (
+  proposals: TransactionsProposedEvent[],
+  params: MonitoringParams
+): Promise<TransactionsProposedEvent[]> => {
+  // Get all assertion Ids from executed proposals covering modules in input proposals.
+  const executedAssertionIds = (
+    await Promise.all(
+      Array.from(new Set(proposals.map((proposal) => proposal.address))).map(async (ogAddress) => {
+        const og = await getOgByAddress(params, ogAddress);
+        const executedProposals = await runQueryFilter<ProposalExecutedEvent>(og, og.filters.ProposalExecuted(), {
+          start: 0,
+          end: params.blockRange.end,
+        });
+        return executedProposals.map((executedProposal) => executedProposal.args.assertionId);
+      })
+    )
+  ).flat();
+
+  // Filter out all proposals that have been executed based on matching assertionId.
+  return proposals.filter((proposal) => !executedAssertionIds.includes(proposal.args.assertionId));
```

Comment:
nit: not a huge deal as long as there are relatively few assertions, but I would suggest avoiding an O(n^2) operation here by putting executedAssertionIds into a Set and checking .has() rather than includes() on an array, which has to loop through the entire array.

Reply:
Agree, I'll do this optimization in a separate PR as there are also other parts of oSnap bot where this can be improved